sysfs/class_sas_device.go

```diff
@@ -0,0 +1,174 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build linux
+// +build linux
+
+package sysfs
+
+import (
+	"github.com/prometheus/procfs/internal/util"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+)
+
+const sasDeviceClassPath = "class/sas_device"
+const sasEndDeviceClassPath = "class/sas_end_device"
+const sasExpanderClassPath = "class/sas_expander"
+
+type SASDevice struct {
+	Name         string   // /sys/class/sas_device/<Name>
+	SASAddress   string   // /sys/class/sas_device/<Name>/sas_address
+	SASPhys      []string // /sys/class/sas_device/<Name>/device/phy-*
+	SASPorts     []string // /sys/class/sas_device/<Name>/device/ports-*
+	BlockDevices []string // /sys/class/sas_device/<Name>/device/target*/*/block/*
+}
+
+type SASDeviceClass map[string]SASDevice
+
+// sasDeviceClasses reads all of the SAS devices from a specific set
+// of /sys/class/sas*/ entries.  The sas_device, sas_end_device, and
+// sas_expander classes are all nearly identical and can be handled by the same basic code.
+
+func (fs FS) parseSASDeviceClass(dir string) (SASDeviceClass, error) {
+	path := fs.sys.Path(dir)
+
+	dirs, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	sdc := make(SASDeviceClass, len(dirs))
+
+	for _, d := range dirs {
+		device, err := fs.parseSASDevice(d.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		sdc[device.Name] = *device
+	}
+
+	return sdc, nil
+}
+
+// SASDeviceClass parses devices in /sys/class/sas_device.
+func (fs FS) SASDeviceClass() (SASDeviceClass, error) {
+	return fs.parseSASDeviceClass(sasDeviceClassPath)
+}
+
+// SASEndDeviceClass parses devices in /sys/class/sas_end_device.
+// This is a subset of sas_device, and excludes expanders.
+func (fs FS) SASEndDeviceClass() (SASDeviceClass, error) {
+	return fs.parseSASDeviceClass(sasEndDeviceClassPath)
+}
+
+// SASExpanderClass parses devices in /sys/class/sas_expander.
+// This is a subset of sas_device, but only includes expanders.
+func (fs FS) SASExpanderClass() (SASDeviceClass, error) {
+	return fs.parseSASDeviceClass(sasExpanderClassPath)
+}
+
+// Parse a single sas_device.  This uses /sys/class/sas_device, as
+// it's a superset of the other two directories so there's no reason
+// to plumb the path through to here.
+func (fs FS) parseSASDevice(name string) (*SASDevice, error) {
+	device := SASDevice{Name: name}
+
+	devicepath := fs.sys.Path(filepath.Join(sasDeviceClassPath, name, "device"))
+
+	dirs, err := ioutil.ReadDir(devicepath)
+	if err != nil {
+		return nil, err
+	}
+
+	phyDevice := regexp.MustCompile(`^phy-[0-9:]+$`)
```

Comment:
Probably makes sense to make this a global var to avoid recompiling the regex

Reply:
Done (and elsewhere).  I debated moving all of the regexes to either class_sas_head.go or their own file (due to global naming conflicts), but ended up removing duplicates and mostly leaving them in the file where they're first used.